plugin/ipam/driver.go

```diff
@@ -44,10 +61,33 @@ func (i *ipam) ReleasePool(poolID string) error {
 	return nil
 }
 
-func (i *ipam) RequestAddress(poolID string, address net.IP, options map[string]string) (*net.IPNet, map[string]string, error) {
+func (i *ipam) RequestAddress(poolID string, address net.IP, options map[string]string) (ip *net.IPNet, _ map[string]string, err error) {
 	Log.Debugln("RequestAddress", poolID, address, options)
-	// Pass magic string to weave IPAM, which then stores the address under its own string
-	ip, err := i.weave.AllocateIP("_")
+	// If we pass magic string "_" to weave IPAM it stores the address under its own string
+	if poolID == "weavepool" { // old-style
+		ip, err = i.weave.AllocateIP("_")
+	} else if strings.HasPrefix(poolID, "weave-") {
+		dashPos := strings.Index(poolID[6:], "-")
+		if dashPos == -1 {
+			return nil, nil, fmt.Errorf("Unrecognized pool ID: %s", poolID)
+		}
+		subnetStr := poolID[6 : dashPos+6]
+		iprangeStr := poolID[dashPos+7:]
+		var subnet, iprange *net.IPNet
+		if _, subnet, err = net.ParseCIDR(subnetStr); err != nil {
+			return
+		}
+		if _, iprange, err = net.ParseCIDR(iprangeStr); err != nil {
+			return
+		}
+		// We are lying slightly to IPAM here: the range is not a subnet
+		if ip, err = i.weave.AllocateIPInSubnet("_", iprange); err != nil {
+			return
+		}
+		ip.Mask = subnet.Mask // fix up the subnet we lied about
+	} else {
+		err = fmt.Errorf("Unrecognized pool ID: %s", poolID)
+	}
```

Comment:
cosmetic: I'd tried to get that big `else if` block outdented.

It also troubles me that some error returns are logged under debug and some are not. You could ensure that all get logged by moving the return logging into a `defer` at the beginning. And you may want to do the same in `RequestPool`. Or ditch the debug logging of returns.